src/Microsoft.AspNetCore.Authentication.Twitter/TwitterHandler.cs

```diff
@@ -148,53 +149,52 @@ public TwitterHandler(IOptionsMonitor<TwitterOptions> options, ILoggerFactory lo
             await Events.RedirectToAuthorizationEndpoint(redirectContext);
         }
 
-        private async Task<RequestToken> ObtainRequestTokenAsync(string callBackUri, AuthenticationProperties properties)
+        public async Task<HttpResponseMessage> ExecuteRequestAsync(string url, HttpMethod httpMethod, RequestToken accessToken = null, Dictionary<string, string> extraOAuthPairs = null, Dictionary<string, string> queryParameters = null, Dictionary<string, string> formData = null)
         {
-            Logger.ObtainRequestToken();
-
-            var nonce = Guid.NewGuid().ToString("N");
+            var canonicalizedRequestBuilder = new StringBuilder();
+            canonicalizedRequestBuilder.Append(httpMethod.Method);
+            canonicalizedRequestBuilder.Append("&");
+            canonicalizedRequestBuilder.Append(Uri.EscapeDataString(url));
 
-            var authorizationParts = new SortedDictionary<string, string>
+            var authorizationParts = new SortedDictionary<string, string>(extraOAuthPairs ?? new Dictionary<string, string>())
             {
-                { "oauth_callback", callBackUri },
                 { "oauth_consumer_key", Options.ConsumerKey },
-                { "oauth_nonce", nonce },
+                { "oauth_nonce", Guid.NewGuid().ToString("N") },
                 { "oauth_signature_method", "HMAC-SHA1" },
                 { "oauth_timestamp", GenerateTimeStamp() },
                 { "oauth_version", "1.0" }
             };
 
-            var parameterBuilder = new StringBuilder();
-            foreach (var authorizationKey in authorizationParts)
+            if (accessToken != null)
             {
-                parameterBuilder.AppendFormat("{0}={1}&", UrlEncoder.Encode(authorizationKey.Key), UrlEncoder.Encode(authorizationKey.Value));
+                authorizationParts.Add("oauth_token", accessToken.Token);
             }
-            parameterBuilder.Length--;
-            var parameterString = parameterBuilder.ToString();
 
-            var canonicalizedRequestBuilder = new StringBuilder();
-            canonicalizedRequestBuilder.Append(HttpMethod.Post.Method);
-            canonicalizedRequestBuilder.Append("&");
-            canonicalizedRequestBuilder.Append(UrlEncoder.Encode(RequestTokenEndpoint));
+            var signatureParts = new SortedDictionary<string, string>((queryParameters ?? new Dictionary<string, string>()).Union(authorizationParts).Union(formData ?? new Dictionary<string, string>()).ToDictionary(x => x.Key, x => x.Value));
```

Comment:
This is the one linq expression that it may make sense to keep, but it needs a more readable layout rather than stringing it out all on one line.